library/src/engine/engine.ts

```diff
@@ -14,19 +15,96 @@ import {
   type InitContext,
   type OnRemovalFn,
   PluginType,
-  type RemovalEntry,
   Requirement,
   type RuntimeContext,
   type RuntimeExpressionFunction,
   type WatcherPlugin,
 } from './types'
 
+const removalKey = (k: string, v: string) => `${k}${DSP}${v}`
+
 export class Engine {
   #signals = new SignalsRoot()
   #plugins: AttributePlugin[] = []
   #actions: ActionPlugins = {}
   #watchers: WatcherPlugin[] = []
-  #removals = new Map<Element, RemovalEntry>()
+
+  // Map of cleanup functions by element, keyed by the raw dataset key and value
+  #removals = new Map<Element, Map<string, OnRemovalFn>>()
+
+  constructor() {
+    const dsPrefix = 'data-'
+
+    const ob = new MutationObserver((mutations) => {
+      for (const {
+        target,
+        type,
+        attributeName,
+        oldValue,
+        addedNodes,
+        removedNodes,
+      } of mutations) {
+        switch (type) {
+          case 'childList':
+            {
+              if (removedNodes.length) {
```

Comment:
Are these wrapping if statements redundant considering that, if `removedNodes` or `addedNodes` have zero length, the for loop wouldn't iterate over anything?

Reply:
True, it's vestigial from another change that got pulled out